Recommit [RISCV] RISCV vector calling convention (2/2) (#79096)

4vtomat opened the pull request:
Bug fix: Handle RVV return type in calling convention correctly.
Return values are handled in a same way as function arguments.
One thing to mention is that if a type can be broken down into homogeneous
vector types, e.g. {<vscale x 4 x i32>, {<vscale x 4 x i32>, <vscale x 4 x i32>}},
it is considered as a vector tuple type and need to be handled by tuple
type rule.

4vtomat commented:
This pr resolves https://github.com/llvm/llvm-project/issues/87323

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21077,7 +21079,82 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+    // First, extract the first element in the argument type.
+    MVT FirstArgRegType;
+    unsigned FirstArgElements = 0;
+    typename SmallVectorImpl<Arg>::const_iterator It;
```
Maybe more readable to use `auto` like this

```
auto It = ArgList.begin();
for (; It != ArgList.end(); ++It)
```

4vtomat replied:
Ok, seems better coding style~

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21077,7 +21079,82 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+void RVVArgDispatcher::constructArgInfos(ArrayRef<Arg> ArgList) {
+  // This lambda determines whether an array of types are constructed by
+  // homogeneous vector types.
+  auto isHomogeneousScalableVectorType = [&](ArrayRef<Arg> ArgList) {
```
What is the `[&]`? Is skimmed the code, but couldn't find anything. Maybe I missed it.

4vtomat replied:
It should be removed.

wangpc-pp commented on llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:
```diff
@@ -409,7 +409,7 @@ bool RISCVCallLowering::lowerReturnVal(MachineIRBuilder &MIRBuilder,
 
   RVVArgDispatcher Dispatcher{&MF, getTLI<RISCVTargetLowering>(),
-                              F.getReturnType()};
+                              ArrayRef(F.getReturnType())};
```
Do we need to construct an ArrayRef explicitly?

4vtomat replied:
Yeah we do, since `ArrayRef` is also a template class, it can't deduct the type, you can try a simple example:
```
    1 template <typename T>
    2 class abc {
    3 public:
    4   abc(T);
    5 };
    6
    7 template <typename T>
    8 void test(abc<T> a);
    9
   10 int main()
   11 {
   12   test(1);
   13 }
```
the code can't compile since `void test(abc<T> a)` doesn't know that 1 can be implicitly constructed, it only know that 1's type is not equal to `abc` lol~

rofirrim replied:
In this example you could explicitly set T in the call (`test<int>(1)`) but… you can't do that in a constructor call like in your case of `RVVArgDispatcher`, so I'm afraid the explicit cast will have to remain.

4vtomat commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21077,7 +21079,82 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+      if (!It->Flags.isSplit() || It->Flags.isSplitEnd())
+        break;
+    }
+    ++It;
```
Sure, it's a good idea to always check it, thanks!

4vtomat commented on llvm/lib/Target/RISCV/RISCVISelLowering.h:
```diff
@@ -1059,7 +1062,7 @@ class RVVArgDispatcher {
   unsigned CurIdx = 0;
 
-  void constructArgInfos(ArrayRef<Type *> TypeList);
+  template <typename Arg> void constructArgInfos(ArrayRef<Arg> Ret);
```
The reason I used template here is that `Arg == ISD::OutputArg` and `Arg == ISD::InputArg` use exact same code and it doesn't need to copy the code. Or is there any other way I can do so?

rofirrim replied:
Apologies, my bad. I misread the code and I thought it was all explicit specialisations. Ignore me.

rofirrim commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21077,7 +21079,92 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+
+    for (; It != ArgList.end(); ++It) {
+      FirstArgRegType = It->VT;
+      FirstArgRegType.dump();
```
I was testing the patch and I was a bit confused by this :)

4vtomat replied:
Oh, I forgot to remove it lol~

4vtomat committed:
Recommit [RISCV] RISCV vector calling convention (2/2) (#79096)

Return values are handled in a same way as function arguments.
One thing to mention is that if a type can be broken down into homogeneous
vector types, e.g. {<vscale x 4 x i32>, {<vscale x 4 x i32>, <vscale x 4 x i32>}},
it is considered as a vector tuple type and need to be handled by tuple
type rule.

4vtomat commented:
Integrated the **"Recommit [RISCV] RISCV vector calling convention (2/2)"** and bug fix patch.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -21102,6 +21091,206 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
+    for (; It != ArgList.end(); ++It) {
+      ArgRegType = It->VT;
+      ++ArgElements;
+      if ((!It->Flags.isSplit() && !IsPart) || It->Flags.isSplitEnd()) {
```
Are all types in a split guaranteed to be the same type or do we need to check all of the types?

4vtomat replied:
> Do we have tests for the split case? That should only happen for types larger than LMUL=8 I think? Do we even want to consider those as tuple types?

You are right and I think the split case doesn't need to be considered as tuple type since we only have 16 vregs, and the minimum split type is also LMUL=16, so it's impossible that argument that has more than 1 split type can be fully passed by reg.
I guess it can be handled as normal vector type.

topperc reviewed (approved):
LGTM

topperc reviewed (approved):
LGTM
